Sprint 6 solution

nev-ivan opened the pull request.

Commit:
Первая итерация 6-го ТЗ.
- жду замечаний, хорошего дня!

Commit:
Первая итерация 6-го ТЗ.
- жду замечаний, хорошего дня!

Commit:
Merge remote-tracking branch 'origin/sprint_6-solution' into sprint_6-solution

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -3,26 +3,91 @@
+    private ArrayList<Task> getTasks() {
+        ArrayList<Task> tasks = new ArrayList<>();
+        Node node = first;
+        do {
+            tasks.add(node.task);
+            node = node.next;
+        } while (node != null);
```
А если история пустая? Тогда мы упадем c NPE, тут подойдет обычный while

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -3,26 +3,91 @@
+
+    private void linkLast(Task task) {
+        final Node node = new Node(task, null, null);
+        if (nodeMap.containsKey(task.getId())) {
+            Node oldNode = nodeMap.remove(task.getId());
+            removeNode(oldNode);
+        }
```
Это все-таки реализация добавления ноды, удаление дубликата логичнее вызвать из метода `add()`, при этом проверку на наличие ноды можно сделать в методе `removeNode()`

Приватные методы обычно размещают после публичных

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -3,26 +3,91 @@
-            historyView.removeFirst();
+    public void remove(int id) {
+        if (nodeMap.containsKey(id)) {
+            final Node node = nodeMap.remove(id);
```
Удаление из хешмапы и проверку на наличие ноды в списке лучше сделать в `removeNode()` методе

nev-ivan replied:
Вопрос, почему удаление из мапы нужно реализовывать в removeNode(), если туда в качестве аргумента передается объект типа Node, а не Task с id, для удобного поиска по мапе? пожалуйста, объясните в следующем ревью

Ivanov-s-tmn replied:
В методе `removeNode()` тоже лучше принимать айди, так будет проще для реализации. А лучше тем, что в методе `removeNode()` будет скрыта вся логика (смена ссылок и удаление из хешмапы). И не нужны будут лишние проверки в методах `add()` и `remove()`, все уже будет в самом методе `removeNode()`

nev-ivan replied:
теперь метод removeNode() принимает id, так действительно удобней, но это отклонение от ТЗ, не страшно?

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -29,44 +34,50 @@ public ArrayList<EpicTask> getEpicTasks() {
 
     @Override
     public void deleteAllTasks() {
+        for (Map.Entry<Integer, Task> taskEntry : taskMap.entrySet()) {
```
Можно просто обойти ключи `keySet()`, тут нет необходимости в обходе полного `Map.Entry`

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -29,44 +34,50 @@ public ArrayList<EpicTask> getEpicTasks() {
 
     @Override
     public void deleteAllEpic() {
+        deleteAllSubTasks();
```
На самом деле в `deleteAllSubTasks()` есть лишняя логика по удалению сабтасок из списка внутри эпика и по обновлению статуса

nev-ivan replied:
то есть конкретно в методе deleteAllEpic() стоит вместо встроенного метода deleteAllSubTasks() прописать код по удалению subTask без обновления статусов и очистки списков?

Ivanov-s-tmn replied:
Да лучше так, в метод `deleteAllSubTasks()` может быть добавлена еще дополнительная логика, которая будет не нужна в `deleteAllEpic()`

Ivanov-s-tmn commented on test/manager/InMemoryHistoryManagerTest.java:
В тестах истории нужно проверить удаление из разных частей истории (из начала, из середины и с конца), важно убедиться, что после удаления из определенной части истории порядок задач корректный

Commit:
Вторая итерация 6-го ТЗ.
- Добавить тесты по удалению задач из истории просмотров
(разные сценарии расположение)
- Перенести функционал по поиску ноды и удалению из мапы в метод removeNode()
- В методе getTasks() изменить тип цикла на while
- При удалении всех тасков, сабтасков, епиков итерацию сделать по ключам

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -45,17 +64,21 @@ private void linkLast(Task task) {
     }
 
     private void removeNode(Node node) {
+        if (!nodeMap.containsValue(node) || node == null) {
+            return;
+        }
```
Тут достаточно только проверки `if (node == null)`, будет передан null из методов add() и remove() если нода не будет найдена по айди в хешмапе `nodeMap.get(id);`

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -45,17 +64,21 @@ private void linkLast(Task task) {
         if (node.prev == null && node.next == null) {
             first = null;
             last = null;
+            nodeMap.remove(node.task.getId());
```
Тоже не очень хорошо дублировать, тут можно объединить первое условие с остальными и убрать return

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -34,28 +34,32 @@ public ArrayList<EpicTask> getEpicTasks() {
+        for (Integer key : subTaskMap.keySet()) {
+            historyManager.remove(key);
+        }
+        for (Map.Entry<Integer, EpicTask> epic : epicMap.entrySet()) {
```
Можно обойти только значение как и с ключами, есть метод `values()`

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -34,28 +34,32 @@ public ArrayList<EpicTask> getEpicTasks() {
-        deleteAllSubTasks();
-        for (Map.Entry<Integer, EpicTask> epicTaskEntry : epicMap.entrySet()) {
-            historyManager.remove(epicTaskEntry.getKey());
+        for (Integer key : subTaskMap.keySet()) {
```
Переменные лучше называть более понятными именами, понятно что это ключ и в самом цикле стоит указать, что ключ - `subtaskId`

Ivanov-s-tmn commented on test/manager/InMemoryHistoryManagerTest.java:
```diff
@@ -8,24 +8,68 @@
+        int idOfTask2 = historyManager.getHistory().get(1).getId();
+        int idOfTask3 = historyManager.getHistory().get(2).getId();
+        historyManager.remove(task1.getId());
+        assertEquals(2, historyManager.getHistory().size(), "некорректное значение");
+        assertEquals(idOfTask2, historyManager.getHistory().get(0).getId());
+        assertEquals(idOfTask3, historyManager.getHistory().get(1).getId());
```
Можно проверить одним ассертом:
`assertEquals(List.of(task2, task3), historyManager.getHistory());`

nev-ivan replied:
я так понимаю проверку на размерность лучше тоже убрать? мы ведь и так проверяем равенство листу из 2-х элементов?

Ivanov-s-tmn replied:
Да, такой вариант проверит  списки на полное соответствие.

Commit:
Третья итерация 6-го ТЗ.
- Отредактировать, упростить работу тестов по удалению задач из истории
- Отредактировать метод removeNode(), убрать дублирующиеся строки
- Заменить имена переменных в циклах по перебору id

Commit:
Третья итерация 6-го ТЗ.
- Изменить параметр метода removeNode()
